### DISCUSSION TITLE
cr2/cr3 to TIFF color temperature

### DISCUSSION THREAD

**Discussion by radimspetlik** [ImageMagick/ImageMagick#4277] [Digital Image Processing] [answered]
Hi, I'd like to use IM to develop cr2/cr3 and store them to TIFFs. It would be enough to keep all the settings as were set in camera during the capture.

I tried

`magick -verbose IMAGE.cr2 -set colorspace RGB -colorspace sRGB IMAGE.TIFF`

which produces an image that is quite different than the default JPG image from the camera.

Is there any setting that would instruct IM to use develop parameters as set by camera? I searched the documentation, didn't find anything...

**Comment by snibgo** [comment, marked as answer]
For colour balance, use "-define dng:use-camera-wb=true" before the input file. See https://imagemagick.org/script/defines.php

Your code suggests that you think the output of "IMAGE.cr2" is linear RGB, so you convert it to sRGB. But that isn't true. With your settings, the image is sRGB. Your command applies a power curve to lighten the image. You might prefer "-auto-level -auto-gamma" instead.

For making an result like the out-of-camera (OOC) JPEG: the algorithms are complex and vary by manufacturer and model, and are proprietary. The algorithms are their secret sauce, a major selling point. They will include colour balance, demosaicing, adjustment of tones and saturation both globally and locally, noise reduction, and sharpening.

The algorithms are usually designed to make the image look pretty. This may not be your objective.

**Reply by radimspetlik** [reply]
@snibgo 

The flag "-define dng: use-camera-wb=true" solved the color temperature problem. Still, the image developed by magick is way darker than when developed by Adobe/Affinity/Canon - in those programs there seems to be some default setting which produces results looking similar to what a JPG processed directly by a camera looks like (of course, not exactly, still - very similarly).

I tried the flag "-define dng:no-auto-bright=true" without success... Any ideas?

**Comment by snibgo** [comment]
I expect the Canon software has the same prettifying software as Canon cameras.

Most raw developers contain software that can be used for prettifying. This may include modules that adjust the raw image to deliberately look more like the JPEG. Some developers will default to those actions, so we open a raw file and immediately the result looks like the JPEG. There is disagreement within the image processing community about whether this is a good idea.

Like camera manufacturers, commercial software developers tend towards prettifying actions. Perhaps this helps to sell software.

IM contains all the operations needed for prettifying, and much else. For example, see my (rather old) page [Adding zing to photographs](http://im.snibgo.com/zingph.htm).

But IM won't prettify by default. When processing raw images, by default all it does (via libraw or dcraw) is apply the camera matrix and demosaic the image, and convert it to sRGB.

**Reply by fmw42** [reply]
@snibgo,  What about exporting both the RAW to TIFF image and JPG image and do color matching conversion of the TIFF to match the JPG? You have the color matching script already.  Have you ever tried that to see if it comes close?

**Comment by snibgo** [comment]
Yes, we can easily make a developed image look like the JPEG. For example, we can [match histograms](http://im.snibgo.com/customim.htm#matchhist) to get similar ranges of tones and colours (hue and saturation).

For example, we develop a raw file with no prettifying to make a tiff (all images shrunk to 600x600, and saved to JPG, for upload speed):
![x_sml](https://user-images.githubusercontent.com/33812368/134779913-92f52108-82d2-4e87-a500-e3c5c83ac311.jpg)

The OOC JPEG looks like this:
![jpg_sml](https://user-images.githubusercontent.com/33812368/134779932-e5444fae-af01-40ea-a84f-725402c59c75.jpg)

Making the developed image look like the JPEG:
![x2_sml](https://user-images.githubusercontent.com/33812368/134779954-2790afa8-321a-4b6a-bd32-a628226309f7.jpg)

The result is 2.3% RMSE from the JPEG, which may be close enough.

**Reply by radimspetlik** [reply]
OK, I get it - there is no way to access the setting in which the image was processed in-camera so we need to reverse-engineer the processing pipeline :)

Thanks!

**Comment by snibgo** [comment]
Getting at the settings isn't the problem. Exiftool probably gives us those. The problem is that we don't know exactly what process makes the JPEG, and precisely what those settings mean within that process. Yes, we would need to reverse-engineer the processing pipeline.

If someone wants to reverse-engineer, and has the time, good luck to them. I don't think it is worth the effort. When I manually edit, I can always get a result I prefer to the OOC JPEG. When I process automatically, I get a result that I usually (but not always) prefer to the OOC JPEG.

For me, the JPEG is just one example of an image I can get from the raw.
